index.js

```diff
@@ -1,2 +1,624 @@
+ *   the secret key is associated, e.g. the user's email address.
+ * @param {String} [options.type="totp"] Either "hotp" or "totp".
+ * @param {Integer} [options.counter] The initial counter value, required
+ *   for HOTP.
+ * @param {String} [options.issuer] The provider or service with which the
+ *   secret key is associated.
+ * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
+ *   sha512).
+ * @param {Integer} [options.digits=6] The number of digits for the one-time
+ *   passcode. Currently ignored by Google Authenticator.
+ * @param {Integer} [options.period=30] The length of time for which a TOTP
+ *   code will be valid, in seconds. Currently ignored by Google
+ *   Authenticator.
+ * @param {String} [options.encoding] Key encoding (ascii, hex, base32,
+ *   base64). If the key is not encoded in Base-32, it will be reencoded.
+ * @return {String} A URL suitable for use with the Google Authenticator.
+ * @see https://github.com/google/google-authenticator/wiki/Key-Uri-Format
+ */
+
+exports.otpauthURL = function otpauthURL (options) {
+  // unpack options
+  var secret = options.secret;
+  var label = options.label;
+  var issuer = options.issuer;
+  var type = (options.type || 'totp').toLowerCase();
+  var counter = options.counter;
+  var algorithm = options.algorithm;
+  var digits = options.digits;
+  var period = options.period;
+  var encoding = options.encoding || 'ascii';
+
+  // validate type
+  switch (type) {
+    case 'totp':
+    case 'hotp':
+      break;
+    default:
+      throw new Error('Speakeasy - otpauthURL - Invalid type `' + type + '`; must be `hotp` or `totp`');
+  }
+
+  // validate required options
+  if (!secret) throw new Error('Speakeasy - otpauthURL - Missing secret');
+  if (!label) throw new Error('Speakeasy - otpauthURL - Missing label');
+
+  // require counter for HOTP
+  if (type === 'hotp' && (counter === null || typeof counter === 'undefined')) {
+    throw new Error('Speakeasy - otpauthURL - Missing counter value for HOTP');
+  }
+
+  // convert secret to base32
+  if (encoding !== 'base32') secret = new Buffer(secret, encoding);
+  if (Buffer.isBuffer(secret)) secret = base32.encode(secret);
+
+  // build query while validating
+  var query = {secret: secret};
+  if (issuer) query.issuer = issuer;
+
+  // validate algorithm
+  if (algorithm != null) {
+    switch (algorithm.toUpperCase()) {
+      case 'SHA1':
+      case 'SHA256':
+      case 'SHA512':
+        break;
+      default:
+        throw new Error('Speakeasy - otpauthURL - Invalid algorithm `' + algorithm + '`');
+    }
+    query.algorithm = algorithm.toUpperCase();
+  }
+
+  // validate digits
+  if (digits != null) {
+    switch (parseInt(digits, 10)) {
+      case 6:
+      case 8:
+        break;
+      default:
+        throw new Error('Speakeasy - otpauthURL - Invalid digits `' + digits + '`');
+    }
+    query.digits = digits;
+  }
```

Comment:
No reason other than Google, but we could just make a warning instead of full-blown error.

> On Jan 26, 2016, at 8:51 PM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js <https://github.com/speakeasyjs/speakeasy/pull/44#discussion_r50941080>:
> 
> > +        throw new Error('Speakeasy - otpauthURL - Invalid algorithm `' + algorithm + '`');
> > +    }
> > +    query.algorithm = algorithm.toUpperCase();
> > +  }
> > +
> > +  // validate digits
> > +  if (digits != null) {
> > +    switch (parseInt(digits, 10)) {
> > +      case 6:
> > +      case 8:
> > +        break;
> > +      default:
> > +        throw new Error('Speakeasy - otpauthURL - Invalid digits `' + digits + '`');
> > +    }
> > +    query.digits = digits;
> > +  }
> @mikepb <https://github.com/mikepb>, do you think we need to be restricting the digits here if it's a generic otpauth:// generator? I haven't been able to find any sort of spec for otpauth:// other than the one in the Google Authenticator wiki.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/speakeasyjs/speakeasy/pull/44/files#r50941080>.
>